### PR TITLE
[WIP] Add ability to generate insert with returning statement

### PR DESCRIPTION
**Pull request by solnic**
This is a spike but I'm opening a PR early to get some input.

I also added key attributes to veritas [here](https://github.com/solnic/veritas/tree/key-attributes) and somehow make the inserts working in veritas-do-adapter [here](https://github.com/solnic/veritas-do-adapter/tree/writes).

I think it would be good to extend `Tuple` so it can return all attributes, only key attributes or all non-key attributes. This would make it a bit easier.

I'm also not sure if the way I change insertion's left and right on-the-fly is a good idea.

Basically I did the simplest thing that could possibly work :)

### CODE REVIEW

File: lib/veritas/sql/generator/relation.rb [hunk 1]
```diff
@@ -165,8 +165,8 @@ def explicit_columns
         # @return [#to_s]
         #
         # @api private
-        def column_list_for(columns)
-          sql = columns.values_at(*@header)
+        def column_list_for(columns, header = nil)
+          sql = columns.values_at(*(header || @header))
```
Comment (dkubb) [inline lib/veritas/sql/generator/relation.rb]: What about:

``` ruby
def column_list_for(columns, header = @header)
  sql = columns.values_at(*header)

  # ...
```

File: spec/unit/veritas/sql/generator/relation/insertion/visit_veritas_relation_operation_insertion_spec.rb [hunk 2]
```diff
@@ -15,6 +15,17 @@
   let(:name)          { Attribute::String.new(:name)                     }
   let(:age)           { Attribute::Integer.new(:age, :required => false) }
 
+  context 'inserting a non-empty materialized relation and returning key attributes' do
+    let(:other) { operand.materialize                       }
+    let(:id)    { Attribute::Integer.new(:id, :key => true) }
+    let(:body)  { [ [ nil, 'Dan Kubb', 36 ] ].each               }
```
Comment (dkubb) [inline spec/unit/veritas/sql/generator/relation/insertion/visit_veritas_relation_operation_insertion_spec.rb]: Alignment is off here ;)

**Comment by dkubb** [comment]
@solnic One thing I'd like to add to Veritas is the ability to specify unique indexes and which attributes are part of them. I'd need to have a way to arbitrarily declare one of those as primary, which I guess is what we'd use for RETURNING.

If the header could store this information, and provide a way to return the attributes for the primary key, then it would be simple to have a method that returns a new Tuple with just those attributes.

**Comment by dkubb** [comment]
> I'm also not sure if the way I change insertion's left and right on-the-fly is a good idea.

I don't see this, can you point it out to me?

**Comment by blambeau** [comment]
@dkubb please call that feature "candidate keys" (logical) and not "unique indexes" (physical) ;-) just kidding.

By the way, I looked at veritas-sql-generator the other day and noticed that it always issues a DISTINCT clause when a projection is used, which is sound. In my own investigation with Alf, I was very unpleased to observe that PostgreSQL does not apply key inference rules _at all_. So for example, observe the following query plans:

```
athena:pointguard blambeau$ psql pointguard_test
psql (9.1.3, server 9.1.2)
Type "help" for help.

pointguard_test=> explain select user_id from auth_users;
                        QUERY PLAN                        
----------------------------------------------------------
 Seq Scan on auth_users  (cost=0.00..1.05 rows=5 width=4)
(1 row)

pointguard_test=> explain select distinct user_id from auth_users;
                           QUERY PLAN                           
----------------------------------------------------------------
 HashAggregate  (cost=1.06..1.11 rows=5 width=4)
   ->  Seq Scan on auth_users  (cost=0.00..1.05 rows=5 width=4)
(2 rows)
```

which means that, in practice, implementing candidate key inference rules in tools like Veritas is probably needed. I've done it in Alf recently. I was rather simple and gives good results. The only rule that I'm currently missing is one that would be smart enough to detect the following case:

```
supplies = ... # relation with a candidate key on [:sid, :pid]
project(restrict(supplies, ->{ pid == 'P5' }), [ :sid ])
# => SELECT sid FROM supplies WHERE pid='P5'
```

The resulting relation has a candidate key on `sid` because `pid` is a constant and can thus by safely projected away.

**Comment by dkubb** [comment]
@blambeau right, I was probably going to call it "keys", but thanks for the reminder.

Once I have keys and key inference baked into veritas, it should open up some nice things in the SQL generator and the optimizer. On the SQL generator site I can then only use DISTINCT when it's needed, and otherwise omit it for keys like you suggest.

In general my approach has been to use the optimizer to get the internal representation to the simplest possible state, and then combine it with an SQL generator that tries to generate the simplest correct SQL it can. Improvements to the optimizer benefit all backends, which is why I focus there first, but I've found lots of things that can be simplified on the SQL side too.

File: lib/veritas/sql/generator/relation/insertion.rb [hunk 3]
```diff
@@ -46,6 +49,29 @@ def column_list
...
+              util   = self.class
+              left   = normalized.left
+              right  = normalized.right
+              header = left.header.reject(&:key?)
+
+              @left  = util.visit(left.class.new(left.name, header))
+              @right = util.visit(right.project(header).materialize)
+            else
+              super
+            end
```
Comment (solnic) [inline lib/veritas/sql/generator/relation/insertion.rb:66]: @dkubb here's where I 're-write' left and right. This was the only way I could come up with to get correct operands that don't include key attributes.

**Comment by solnic** [comment]
@dkubb re: unique indexes / keys - so a tuple would have a method `#keys` that returns all the keys and also `#primary_key` (or is it a bad name?) which returns the primary key (that may consist of multiple key attributes of course).

I'm on it :)
